[Bug]: Shutdown hook in RyukResourceReaper prevents a graceful shutdown with spring framework

### Module

Core

### Testcontainers version

1.19.7

### Using the latest Testcontainers version?

Yes

### Host OS

Linux, Windows

### Host Arch

x86_64

### Docker version

```shell
Client:
 Cloud integration: v1.0.35+desktop.13
 Version:           26.0.0
 API version:       1.45
 Go version:        go1.21.8
 Git commit:        2ae903e
 Built:             Wed Mar 20 15:18:56 2024
 OS/Arch:           windows/amd64
 Context:           default

Server: Docker Desktop 4.29.0 (145265)
 Engine:
  Version:          26.0.0
  API version:      1.45 (minimum version 1.24)
  Go version:       go1.21.8
  Git commit:       8b79278
  Built:            Wed Mar 20 15:18:01 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.28
  GitCommit:        ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```


### What happened?

Dears,

With the introduction of this feature: #7717 

The ryuk instance is terminated by a shutdown hook, therefore all containers are closed very early.

This prevents other shutdown hooks (eg: those set by the spring framework) to perform actions needed for a graceful shutdown that still require containers to be up.
Due to a retry mechanism in Spring's shutdown hook, it greatly increases the overall time to run tests!

Could you please revert this change or introduce a mean to disable this shutdown hook, thanks!

### Relevant log output

_No response_

### Additional Information

_No response_

@jobayle I'm not sure what is the reason. The containers are definitely terminated when shutdown listener is executed because I see errors like this in the logs:

<details>
  <summary>Logs</summary>

```
2024-11-01T00:08:43.749+01:00 ERROR 28940 --- [ionShutdownHook] o.s.t.i.TransactionInterceptor           : Application exception overridden by rollback exception

org.springframework.dao.DataAccessResourceFailureException: PreparedStatementCallback; SQL [UPDATE machine SET running_status = ?, last_update = now() WHERE name = ?]; An I/O error occurred while sending to the backend.
        at org.springframework.jdbc.support.SQLStateSQLExceptionTranslator.doTranslate(SQLStateSQLExceptionTranslator.java:121) ~[spring-jdbc-6.1.3.jar:6.1.3]

...

        at bug.machine.MachineRepository.updateRunningStatus(MachineRepository.kt:73) ~[main/:na]
 
...

        at bug.machine.MachineRepository$$SpringCGLIB$$0.updateRunningStatus(<generated>) ~[main/:na]
        at bug.machine.ApplicationStatus.handleShutdown(ApplicationStatus.kt:44) ~[main/:na]

...

        at org.springframework.context.support.AbstractApplicationContext.doClose(AbstractApplicationContext.java:1127) ~[spring-context-6.1.3.jar:6.1.3]

...

Caused by: org.postgresql.util.PSQLException: An I/O error occurred while sending to the backend.
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:398) ~[postgresql-42.7.3.jar:42.7.3]

...

        at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java) ~[HikariCP-5.0.1.jar:na]
        at org.springframework.jdbc.core.JdbcTemplate.lambda$update$2(JdbcTemplate.java:975) ~[spring-jdbc-6.1.3.jar:6.1.3]
        at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:658) ~[spring-jdbc-6.1.3.jar:6.1.3]
        ... 42 common frames omitted
Caused by: java.io.EOFException: null
        at org.postgresql.core.PGStream.receiveChar(PGStream.java:469) ~[postgresql-42.7.3.jar:42.7.3]
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2166) ~[postgresql-42.7.3.jar:42.7.3]
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:371) ~[postgresql-42.7.3.jar:42.7.3]
        ... 50 common frames omitted

```
</details>

I think it could be ordering and speed of shutdown of different systems issue, yes. As we see, it is `java.io.EOFException` in my case. It looks like in my case Hikari already tested the connection before returning it from the pool and the connection was valid, but then, when context's listener tried to execute the `UPDATE` query, the Postgres container was finally stopped right in the middle of the query, that is why Hikari doesn't try to reconnect and timeout doesn't work. The exception is happening when execution is already in the user code. To make Hikari timeout work, the exception should happen earlier, when Hikari tests the connection before returning it from the pool to the user code. So, probably, my machine is too slow on stopping docker containers. :)

**Update:**
I actually just made your example reproducible by adding 1s delay to the `ContextClosedEvent` listener:

```java
    @EventListener
    fun handleShutdown(ctxStatus: ContextClosedEvent) {
        Thread.sleep(1000)
        logger.info("UPDATING MACHINE CONFIGURATION STATUS=SHUTDOWN")
        machineRepository.updateRunningStatus(machineConfiguration.name, RunningStatus.OFFLINE)
    }
```

This way the tests are executed solid 1m 57s. Yay! 😂 

With many

```
HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@3eeeae1f (This connection has been closed.). Possibly consider using a shorter maxLifetime value.
```

and

```
reconnecting RedisConnection@618341304 [redisClient=[addr=redis://localhost:51852], channel=[id: 0x145b0193, L:/127.0.0.1:51949 ! R:localhost/127.0.0.1:51852], currentCommand=null, usage=0] to localhost/127.0.0.1:51852
```

--- a/core/src/main/java/org/testcontainers/utility/RyukResourceReaper.java
+++ b/core/src/main/java/org/testcontainers/utility/RyukResourceReaper.java
@@ -77,19 +77,6 @@ class RyukResourceReaper extends ResourceReaper {
 
         ryukContainer.start();
 
-        Runtime
-            .getRuntime()
-            .addShutdownHook(
-                new Thread(
-                    DockerClientFactory.TESTCONTAINERS_THREAD_GROUP,
-                    () -> {
-                        this.dockerClient.killContainerCmd(this.ryukContainer.getContainerId())
-                            .withSignal("SIGTERM")
-                            .exec();
-                    }
-                )
-            );
-
         CountDownLatch ryukScheduledLatch = new CountDownLatch(1);
 
         String host = ryukContainer.getHost();

